File: 6. Count nodes.cpp
/*
 Task: Count nodes
 Time Complexity: O(n)
 Space Complexity: O(h) — recursion stack
*/

#include <bits/stdc++.h>

using namespace std;

struct Node
{
    int data;
    Node* left;
    Node* right;

    Node(int d) : data(d), left(nullptr), right(nullptr) {}
};

int countNodes(Node* root)
{
    if (!root)
    {
        return 0;
    }

    return 1 + countNodes(root->left) + countNodes(root->right);
}

int main()
{
    Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);

    cout << "Count: " << countNodes(root) << endl;
    return 0;
}

// sc tc

File: 6. Count nodes.py
"""
Task: Count nodes in binary tree
Time Complexity: O(n)
Space Complexity: O(h) — recursion stack
"""

class Node:
    def __init__(self,data): self.data=data; self.left=None; self.right=None

def count_nodes(root):
    if not root: return 0
    return 1+count_nodes(root.left)+count_nodes(root.right)

if __name__=='__main__':
    root=Node(1); root.left=Node(2); root.right=Node(3); root.left.left=Node(4)
    print('Count:', count_nodes(root))
